executors/accelerate/src/hypha/accelerate_executor/training.py

```diff
@@ -64,98 +83,185 @@ def main(socket_path: str, work_dir: str, job_json: str) -> None:  # noqa: PLR09
-                        else:
-                            counter -= 1
-                        start_time = datetime.datetime.now()
+        while True:
+            loop_start_ms = time.time() * 1000.0
+            action_resp = session.send_action({"job_id": job_id, "status": current_status})
+            next_action = action_resp.get("next", {})
+
+            if next_action.get("executor") != "train":
+                raise RuntimeError(f"Unexpected executor action: {next_action}")
+
+            action = next_action.get("action", {})
+            kind = action.get("kind")
+
+            print(f"Action: {kind}", flush=True)
 
+            if kind == "terminate":
+                print("Training finished", flush=True)
+                break
+
+            if kind == "idle":
+                timeout_ms = system_time_to_epoch_ms(action.get("timeout"))
+                if timeout_ms is not None:
+                    sleep_until_epoch_ms(timeout_ms)
+                current_status = {"executor": "train", "details": {"state": "idle"}}
+            elif kind == "execute-batch":
+                batch = next(training_data_iter)
+                optimizer.zero_grad()
+                outputs = model(**batch)
+                loss = outputs if isinstance(outputs, torch.Tensor) else outputs["loss"]
+                accelerator.backward(loss)
+                optimizer.step()
+                scheduler.step()
                 if accelerator.is_main_process:
-                    session.send_status("update")
-                    # For testing purposes set to global!
+                    batch_size = next(iter(batch.values())).shape[0]
+                    current_status = {
+                        "executor": "train",
+                        "details": {"state": "batch-completed", "batch_size": batch_size},
+                    }
+                    # Prepare gradients for potential SendUpdate
                     file_name = f"{epoch_counter}_local_gradients.pt"
                     result_path = os.path.join(work_dir, file_name)
-                    # Save unwrapped model to avoid accelerator wrappers interfering
-                    model = accelerator.unwrap_model(model)
-                    # All weights need to be on CPU
-                    model.to("cpu")
+                    # Copy weights to CPU without moving the live model off-device.
+                    model_state = accelerator.unwrap_model(model).state_dict()
+                    state_cpu = {k: v.detach().cpu() for k, v in model_state.items()}
+                    save_file(extract_gradients(state_cpu, previous_model_path), result_path)
+                    last_gradient = file_name
+                    last_metrics = {"loss": float(loss.detach().cpu().numpy())}
+                else:
+                    current_status = {
+                        "executor": "train",
+                        "details": {"state": "batch-completed", "batch_size": 0},
+                    }
```

Comment:
I am pretty sure this will fail with a multi GPU machine. But lets fix this later.